File: scala/parsers/src/main/scala/jp/co/teruuu/parser/json/JOrParser.scala
package jp.co.teruuu.parser.json

import jp.co.teruuu.parser.common.{ParseNg, ParseOk, ParseResult}

class JOrParser[T <: JValue,U <: JValue](parser1: JParser[T], parser2: JParser[U]) extends JParser[JValue] {

  override def parse(input: String): ParseResult[JValue] = {
    parser1.parse(input) match {
      case result1: ParseOk[T] => ParseOk(result1.value, result1.next)
      case result1: ParseNg[_] => {
        parser2.parse(input) match {
          case result2: ParseOk[U] => ParseOk(result2.value, result2.next)
          case result2: ParseNg[_] => ParseNg(result2.message, input)
        }
      }
    }
  }
}

File: java/parsers_java8/src/main/java/jp/co/teruuu/parser/json/JOrParser.java
package jp.co.teruuu.parser.json;

import jp.co.teruuu.parser.common.ParseResult;
import jp.co.teruuu.parser.common.Parser;

public class JOrParser<X> implements JParser<X> {
  private JParser<X> lhs;
  private Parser<X> rhs;
  public JOrParser(JParser<X> lhs, JParser<X> rhs) {
    this.lhs = lhs;
    this.rhs = rhs;
  }

  @Override
  public ParseResult<X> parse(String input) {
    ParseResult<X> lresult = lhs.parse(input);
    if(lresult instanceof ParseResult.Failure<?>) {
      return rhs.parse(input);
    } else {
      return lresult;
    }
  }
}
